### ISSUE TITLE
[Bug] `AssertionError: <class 'types.GenericAlias'> is class.` (I don't even know anymore.) :face_exhaling: 

### ISSUE THREAD

**Issue by sylvorg** [beartype/beartype#358]
Hello!

With the following test, adapted from [here](https://github.com/python/cpython/blob/main/Lib/_collections_abc.py#L279):

```python
from abc import ABCMeta
from beartype import beartype
from rich.traceback import install
from types import GenericAlias

install(show_locals=True)

@beartype
class Test(metaclass=ABCMeta):
    __slots__ = ()
    @classmethod
    def __subclasshook__(cls, C: type):
        if cls is Test:
            return True
        return NotImplemented
    __class_getitem__ = classmethod(GenericAlias)
```

I get the following (rich.)traceback:

```python
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ in <module>:1                                                                                    │
│ ╭─────────────────────────────────────── locals ────────────────────────────────────────╮        │
│ │    _mk_param = <function _mk_param at 0x0000024EC56DFEC0>                             │        │
│ │      ABCMeta = <class 'abc.ABCMeta'>                                                  │        │
│ │     beartype = <function beartype at 0x0000024EC5B4CFE0>                              │        │
│ │        etest = <function etest at 0x0000024EC56DFF60>                                 │        │
│ │ GenericAlias = <class 'types.GenericAlias'>                                           │        │
│ │      inspect = <module 'inspect' from 'C:\\Users\\jeetr\\anaconda3\\Lib\\inspect.py'> │        │
│ │      install = <function install at 0x0000024EC614A3E0>                               │        │
│ │          sig = <Signature (a, /, b, c=3, *args, d, e=5, **kwargs)>                    │        │
│ │         test = <function test at 0x0000024EC56DFCE0>                                  │        │
│ ╰───────────────────────────────────────────────────────────────────────────────────────╯        │
│                                                                                                  │
│ C:\Users\jeetr\anaconda3\Lib\site-packages\beartype\_decor\decorcache.py:77 in beartype          │
│                                                                                                  │
│    74 │   # here; callers that are doing this are sufficiently intelligent to be                 │
│    75 │   # trusted to violate PEP 561-compliance if they so choose. So... *shrug*               │
│    76 │   elif obj is not None:                                                                  │
│ ❱  77 │   │   return beartype_object(obj, conf)                                                  │
│    78 │   # Else, we were passed *NO* object to be decorated. In this case, this                 │
│    79 │   # decorator is in configuration rather than decoration mode.                           │
│    80                                                                                            │
│                                                                                                  │
│ ╭──────────── locals ────────────╮                                                               │
│ │ conf = BeartypeConf()          │                                                               │
│ │  obj = <class '__main__.Test'> │                                                               │
│ ╰────────────────────────────────╯                                                               │
│                                                                                                  │
│ C:\Users\jeetr\anaconda3\Lib\site-packages\beartype\_decor\decorcore.py:83 in beartype_object    │
│                                                                                                  │
│    80 │                                                                                          │
│    81 │   # Return either...                                                                     │
│    82 │   return (                                                                               │
│ ❱  83 │   │   _beartype_object_fatal(obj, conf=conf, **kwargs)                                   │
│    84 │   │   # If this beartype configuration requests that this decorator raise                │
│    85 │   │   # fatal exceptions at decoration time, defer to the lower-level                    │
│    86 │   │   # decorator doing so;                                                              │
│                                                                                                  │
│ ╭───────────── locals ─────────────╮                                                             │
│ │   conf = BeartypeConf()          │                                                             │
│ │ kwargs = {}                      │                                                             │
│ │    obj = <class '__main__.Test'> │                                                             │
│ ╰──────────────────────────────────╯                                                             │
│                                                                                                  │
│ C:\Users\jeetr\anaconda3\Lib\site-packages\beartype\_decor\decorcore.py:128 in                   │
│ _beartype_object_fatal                                                                           │
│                                                                                                  │
│   125 │   # Return either...                                                                     │
│   126 │   return (                                                                               │
│   127 │   │   # If this object is a class, this class decorated with type-checking.              │
│ ❱ 128 │   │   beartype_type(obj, **kwargs)  # type: ignore[return-value]                         │
│   129 │   │   if isinstance(obj, type) else                                                      │
│   130 │   │   # Else, this object is a non-class. In this case, this non-class                   │
│   131 │   │   # decorated with type-checking.                                                    │
│                                                                                                  │
│ ╭───────────── locals ──────────────╮                                                            │
│ │ kwargs = {'conf': BeartypeConf()} │                                                            │
│ │    obj = <class '__main__.Test'>  │                                                            │
│ ╰───────────────────────────────────╯                                                            │
│                                                                                                  │
│ C:\Users\jeetr\anaconda3\Lib\site-packages\beartype\_decor\_decortype.py:266 in beartype_type    │
│                                                                                                  │
│   263 │   │   ):                                                                                 │
│   264 │   │   │   # This attribute decorated with type-checking configured by this               │
│   265 │   │   │   # configuration if *NOT* already decorated.                                    │
│ ❱ 266 │   │   │   attr_value_beartyped = beartype_object(                                        │
│   267 │   │   │   │   obj=attr_value,                                                            │
│   268 │   │   │   │   conf=conf,                                                                 │
│   269 │   │   │   │   cls_stack=cls_stack,                                                       │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │            attr_name = '__class_getitem__'                                                   │ │
│ │           attr_value = <classmethod(<class 'types.GenericAlias'>)>                           │ │
│ │ attr_value_beartyped = <classmethod(<function Test.__subclasshook__ at 0x0000024EC6569120>)> │ │
│ │      beartype_object = <function beartype_object at 0x0000024EC5B4D1C0>                      │ │
│ │                  cls = <class '__main__.Test'>                                               │ │
│ │       cls_sizeof_old = <method '__sizeof__' of 'object' objects>                             │ │
│ │            cls_stack = (<class '__main__.Test'>,)                                            │ │
│ │                 conf = BeartypeConf()                                                        │ │
│ │ is_attr_beartypeable = True                                                                  │ │
│ │     is_cls_beartyped = None                                                                  │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ C:\Users\jeetr\anaconda3\Lib\site-packages\beartype\_decor\decorcore.py:83 in beartype_object    │
│                                                                                                  │
│    80 │                                                                                          │
│    81 │   # Return either...                                                                     │
│    82 │   return (                                                                               │
│ ❱  83 │   │   _beartype_object_fatal(obj, conf=conf, **kwargs)                                   │
│    84 │   │   # If this beartype configuration requests that this decorator raise                │
│    85 │   │   # fatal exceptions at decoration time, defer to the lower-level                    │
│    86 │   │   # decorator doing so;                                                              │
│                                                                                                  │
│ ╭─────────────────────── locals ───────────────────────╮                                         │
│ │   conf = BeartypeConf()                              │                                         │
│ │ kwargs = {'cls_stack': (<class '__main__.Test'>,)}   │                                         │
│ │    obj = <classmethod(<class 'types.GenericAlias'>)> │                                         │
│ ╰──────────────────────────────────────────────────────╯                                         │
│                                                                                                  │
│ C:\Users\jeetr\anaconda3\Lib\site-packages\beartype\_decor\decorcore.py:132 in                   │
│ _beartype_object_fatal                                                                           │
│                                                                                                  │
│   129 │   │   if isinstance(obj, type) else                                                      │
│   130 │   │   # Else, this object is a non-class. In this case, this non-class                   │
│   131 │   │   # decorated with type-checking.                                                    │
│ ❱ 132 │   │   beartype_nontype(obj, **kwargs)  # type: ignore[return-value]                      │
│   133 │   )                                                                                      │
│   134                                                                                            │
│   135                                                                                            │
│                                                                                                  │
│ ╭────────────────────────────────── locals ──────────────────────────────────╮                   │
│ │ kwargs = {'conf': BeartypeConf(), 'cls_stack': (<class '__main__.Test'>,)} │                   │
│ │    obj = <classmethod(<class 'types.GenericAlias'>)>                       │                   │
│ ╰────────────────────────────────────────────────────────────────────────────╯                   │
│                                                                                                  │
│ C:\Users\jeetr\anaconda3\Lib\site-packages\beartype\_decor\_decornontype.py:110 in               │
│ beartype_nontype                                                                                 │
│                                                                                                  │
│   107 │   #   * Descriptors created by @staticmethod are technically callable but                │
│   108 │   #     C-based and thus unsuitable for decoration.                                      │
│   109 │   if obj_type in MethodDecoratorBuiltinTypes:                                            │
│ ❱ 110 │   │   return beartype_descriptor_decorator_builtin(obj, **kwargs)  # type: ignore[retu   │
│   111 │   # Else, this object is *NOT* an uncallable builtin method descriptor.                  │
│   112 │   #                                                                                      │
│   113 │   # If this object is uncallable, raise an exception.                                    │
│                                                                                                  │
│ ╭─────────────────────────────────── locals ───────────────────────────────────╮                 │
│ │   kwargs = {'conf': BeartypeConf(), 'cls_stack': (<class '__main__.Test'>,)} │                 │
│ │      obj = <classmethod(<class 'types.GenericAlias'>)>                       │                 │
│ │ obj_type = <class 'classmethod'>                                             │                 │
│ ╰──────────────────────────────────────────────────────────────────────────────╯                 │
│                                                                                                  │
│ C:\Users\jeetr\anaconda3\Lib\site-packages\beartype\_decor\_decornontype.py:457 in               │
│ beartype_descriptor_decorator_builtin                                                            │
│                                                                                                  │
│   454 │   │   #   * The low-level beartype_func() decorator (which requires the passed           │
│   455 │   │   #     object to be callable, which the descriptor created and returned             │
│   456 │   │   #     by the @property decorator is *NOT*).                                        │
│ ❱ 457 │   │   func_checked = beartype_nontype(descriptor.__func__,  **kwargs)  # type: ignore[   │
│   458 │   │                                                                                      │
│   459 │   │   # Return a new class method descriptor decorating the pure-Python                  │
│   460 │   │   # unbound function wrapped by this descriptor with type-checking,                  │
│                                                                                                  │
│ ╭────────────────────────────────────── locals ───────────────────────────────────────╮          │
│ │      descriptor = <classmethod(<class 'types.GenericAlias'>)>                       │          │
│ │ descriptor_type = <class 'classmethod'>                                             │          │
│ │          kwargs = {'conf': BeartypeConf(), 'cls_stack': (<class '__main__.Test'>,)} │          │
│ ╰─────────────────────────────────────────────────────────────────────────────────────╯          │
│                                                                                                  │
│ C:\Users\jeetr\anaconda3\Lib\site-packages\beartype\_decor\_decornontype.py:74 in                │
│ beartype_nontype                                                                                 │
│                                                                                                  │
│    71 │   '''                                                                                    │
│    72 │                                                                                          │
│    73 │   # Validate that the passed object is *NOT* a class.                                    │
│ ❱  74 │   assert not isinstance(obj, type), f'{repr(obj)} is class.'                             │
│    75 │   # print(f'Decorating non-type {repr(obj)}...')                                         │
│    76 │                                                                                          │
│    77 │   # Type of this object.                                                                 │
│                                                                                                  │
│ ╭────────────────────────────────── locals ──────────────────────────────────╮                   │
│ │ kwargs = {'conf': BeartypeConf(), 'cls_stack': (<class '__main__.Test'>,)} │                   │
│ │    obj = <class 'types.GenericAlias'>                                      │                   │
│ ╰────────────────────────────────────────────────────────────────────────────╯                   │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AssertionError: <class 'types.GenericAlias'> is class.
```

How can I define a new type that can accept parameters when using `beartype`?

Thank you kindly for the help!

**Comment by leycec** [comment]
**Woah.** Thanks so much for that *intense* minimal working example, @sylvorg. Super black magic arises. Yet, @beartype should support super black magic. @beartype is clearly mistaken about the nature of reality here. Bugs ahoy, matey!

Let me dig deep into this. I'll spin up a new @beartype 0.18.3 patch release for you next week. I pinky swear it. :crossed_fingers:

**Comment by sylvorg** [comment]
By all means, take your time! I've got assignments and finals to do as well! 😹

**Comment by leycec** [comment]
**Resolved by ce6cfbe1ffd.** We're *this* close to solving @sylvorg's life problems, people. And @sylvorg is *this* close to passing this semester without passing out from chronic fatigue, people.

This is my cheerleading anthem for @sylvorg:

> "Ganbare! Ganbare! @sylvorg is the best! Forget the rest! Never rest! Pass those tests!"

![](https://media1.tenor.com/m/IqZApAU_d98AAAAC/archen-pokemon.gif)

**Comment by sylvorg** [comment]
Ooh, is [Archen](https://bulbapedia.bulbagarden.net/wiki/Archen_(Pok%C3%A9mon)) helping you cheer me on? 😹 Thank you (both) so much for all the effort you've put in so far! 😻
